ract/art/worm_simulator.py

```diff
@@ -0,0 +1,167 @@
+        self._worms = []
+        for worm in range(number_of_worms):
+            self._worms.append(self._setup_worm())
+        for worm in self._worms:
+            self._frame[worm.head_x()][worm.head_y()][worm.head_z()] = \
+                Worm.WORM_INTENSITY
+
+    @staticmethod
+    def _setup_worm():
+        """ Initialize a worm
+        :return:
+        """
+        worm = Worm(np.array([randint(0, 7), randint(0, 7), randint(0, 7)]),
+                    Worm.DIRECTIONS[randint(0, 5)])
+        return worm
+
+    def _move_worm(self):
+        """ Try to add a section to each worm as follows:
+        1. Make a shuffled list of all moves.
+        2. Check if the current move for the worm is legal and apply if possible
+         and exit.
+        3. If the list of moves has been exhausted raise WormDied.
+        3. Take move from the list of moves and make the current move that move
+        :return:
+        """
+        for worm in self._worms:
+            moves = Worm.DIRECTIONS[:]
+            shuffle(moves)
+            try:
+                while True:
+                    new_move = worm.head() + worm.direction
+                    if self._move_valid(new_move):
+                        worm.add_segment(new_move)
+                        break
+                    worm.direction = moves[0]
+                    del(moves[0])
+            except IndexError:
+                worm.alive = False
+                self._worm_death(worm)
+
+    def _move_valid(self, move):
+        """ Legal moves are within the box and not into a piece of worm.
+        :param move: Proposed move
+        :return boolean:
+        """
+        try:
+            occupied = self._frame[move[0]][move[1]][move[2]]
+            if occupied > MIN_INTENSITY:
+                return False
+            if not all(list(move > -1)):
+                return False
+            return True
+        except IndexError:
+            return False
+
+    def _worm_death(self, worm):
+        """ Animate worm death. Worm flashes brightly on death
+        :param worm:
+        :return:
+        """
+        print("worm dies")
+        for segment in worm.segments:
+            self._frame[segment[0]][segment[1]][segment[2]] = MAX_INTENSITY
+
+    def _worm_birth(self):
+        """ Check if we need to rebirth a worm
+        :return:
+        """
+        for worm in self._worms:
+            if not worm.alive:
+                print("worm lives")
+                for segment in worm.segments:
+                    self._frame[segment[0]][segment[1]][segment[2]] = \
+                        MIN_INTENSITY
+                dead_worm_index = self._worms.index(worm)
+                self._worms[dead_worm_index] = self._setup_worm()
+
+    def next_frame(self):
+        if self._stepper.step():
+            self._worm_birth()
+            self._move_worm()
+            for worm in self._worms:
+                self._frame[worm.head_x()][worm.head_y()][worm.head_z()] = \
```

Comment:
Thinking a bit more -- you could simplify this whole thing using `np.put`:
```python
new_frame = constant_frame()
for worm in worms:
    np.put(new_frame, worm.segments, worm.WORM_INTENSITY)
```
This will write `WORM_INTENSITY` to every position in the worm segments (and numpy will do it efficiently without any Python overhead).

Not sure where it is worth doing right this second, but thought I'd mention the idea.